mount/main.go

```diff
@@ -0,0 +1,70 @@
+//  Copyright 2021 Authors of Cilium
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/cilium/cilium/pkg/mountinfo"
+
+	"golang.org/x/sys/unix"
+)
+
+func mountCgroup(cgroupRoot string) error {
+	cgroupRootStat, err := os.Stat(cgroupRoot)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if err = os.MkdirAll(cgroupRoot, 0755); err != nil {
+				return fmt.Errorf("unable to create cgroup mount directory: %s", err)
+			}
+		} else {
+			return fmt.Errorf("failed to stat the mount path %s: %s", cgroupRoot, err)
+		}
+	} else if !cgroupRootStat.IsDir() {
+		return fmt.Errorf("%s is a file which is not a directory", cgroupRoot)
+	}
+
+	if err = unix.Mount("none", cgroupRoot, "cgroup2", 0, ""); err != nil {
+		return fmt.Errorf("failed to mount %s: %s", cgroupRoot, err)
+	}
+
+	return nil
+}
+
+func main() {
+	if len(os.Args) != 2 {
+		fmt.Fprintf(os.Stderr, "usage: %s <cgroup-root> \n\n", os.Args[0])
+		os.Exit(1)
+	}
+
+	cgroupRoot := os.Args[1]
```

Comment:
why can't we simply call `cgroups.CheckOrMountCgrpFS(cgroupRoot)`?

Reply:
We could; it's more or less the same code. But the logic is faulty so we'll need to first merge - https://github.com/cilium/cilium/pull/15999. I also prefer to avoid duplicating the code.

I made one more change to this function - when the passed path has a mount, but different fs type, we currently return error. But that might not be always true. For example, if cgroup v1 fs is mounted, we can still mount cgroup2 fs there. Given that the auto mount path is set to /run/cilium/cgroupv2, this change isn't necessarily required though.

Reply:
PR #15999 should be merged by then.

>Edit : I'm not sure we want to use the exact same function since it has other dependencies like logging, etc.

it's fine, it will make the logging consistent across cilium binaries.

Reply:
See this comment - https://github.com/cilium/cilium/pull/16815#discussion_r665047791. I'm leaning towards keeping the code executed by this program to as minimal as possible. Let's see what @joestringer has to say as well.

Reply:
Six of one, half dozen of the other. It looks to me like lines 59-68 are trying to achieve the same logic as `cgrpCheckOrMountLocation()` so we may as well have exactly one copy of that logic and make sure it's the way we want it for both use cases.